cico_run_coverage.sh

```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+. cico_setup.sh
+
+COVERAGE="f8osoproxy-coverage"
+PACKAGE_NAME="github.com/containous/traefik"
+
+GOPATH_IN_CONTAINER=/tmp/go
+PACKAGE_PATH=$GOPATH_IN_CONTAINER/src/$PACKAGE_NAME
+
+docker build -t "$COVERAGE" -f Dockerfile.builder .
+
+docker run --detach=true -t \
+    --name="$COVERAGE-run" \
+    -v $(pwd):$PACKAGE_PATH:Z \
+    -u $(id -u $USER):$(id -g $USER) \
+    -e GOPATH=$GOPATH_IN_CONTAINER \
+    -w $PACKAGE_PATH \
+    $COVERAGE
+
+docker exec -t "$COVERAGE-run" bash -ec 'go get github.com/jteeuwen/go-bindata/...'
+docker exec -t "$COVERAGE-run" bash -ec 'go generate'
+docker exec -t "$COVERAGE-run" bash -ec 'go build -o dist/traefik ./cmd/traefik'
+
+docker exec -t "$COVERAGE-run" bash -ec 'go test -v ./middlewares/osio/ -coverprofile coverage.middlewares -covermode=set -timeout 5m'
+docker exec -t "$COVERAGE-run" bash -ec 'go test -v ./provider/osio/ -coverprofile coverage.provider -covermode=set -timeout 5m'
+docker exec -t "$COVERAGE-run" bash -ec 'go test -v ./integration/ -integration -osio'
+
+# Upload coverage to codecov.io
+# -t <upload_token> copy from https://codecov.io/gh/fabric8-services/fabric8-oso-proxy/settings
+bash <(curl -s https://codecov.io/bash) -t 3a135505-4f56-4dce-900e-e451b95601e5
```

Comment:
@aslakknutsen just upload all report and let codecov to figure out to take all files with name coverage* .. fine ??
OR any benefit to do it like f8-wit [link](https://github.com/fabric8-services/fabric8-wit/blob/master/cico_setup.sh#L129)

Reply:
I guess it should be fine... The script is supposed to search for the code coverage reports..
@kwk any reason not to use the native report search?

Reply:
@alexeykazakov IIRC we've used

```
     -f FILE     Target file(s) to upload
                 -f "path/to/file"     only upload this file
                                       skips searching unless provided patterns below

                 -f '!*.bar'           ignore all files at pattern *.bar
                 -f '*.foo'            include all files at pattern *.foo
                 Must use single quotes.
                 This is non-exclusive, use -s "*.foo" to match specific paths.
```

to specify a concrete filename to upload because our tests that produce coverage output write them in different files depending on the coverage mode and to keep them separately:

```Make
# Files that combine package coverages for unit- and integration-tests separately
COV_PATH_UNIT = $(TMP_PATH)/coverage.unit.mode-$(COVERAGE_MODE)
COV_PATH_INTEGRATION = $(TMP_PATH)/coverage.integration.mode-$(COVERAGE_MODE)
COV_PATH_REMOTE = $(TMP_PATH)/coverage.remote.mode-$(COVERAGE_MODE)

# File that stores overall coverge for all packages and unit- integration- and remote-tests
COV_PATH_OVERALL = $(TMP_PATH)/coverage.mode-$(COVERAGE_MODE)
```

Then in the cico script we copy `tmp/coverage.mode-*` to `coverage.txt` and upload that one explicitly. We then also tell codecov to not search for any report:

```
-X search        Disable searching for reports
```

The whole reason for doing it this way was that we wanted to be able to run any coverage producing test locally first and then ask for a rendered version of the coverage report without having to run the whole test again. Therefore we have the `Make` targets `coverage-unit-html` and `coverage-integration-html` that call `@go tool cover -html=$(COV_PATH_UNIT)` for example.